### PR TITLE
add sgd for Logistic Regression

### PR DESCRIPTION
**Pull request by rayeaster**
add stochastic gradient descent for Logistic Regression
reference: http://www.cs.cmu.edu/~wcohen/10-605/sgd-part2.pdf

### CODE REVIEW

**Comment by rayeaster** [comment]
unit test result:

```
Segment
Segment error rate = 5.93%
SGD-Segment
SGD-Segemnt one more epoch...
SGD-Segment error rate = 6.17%

Iris binary
Logistic Regression error = 3
Iris
Logistic Regression error = 3
Iris binary
SGD-Iris Binary one more epoch...
SGD-Iris Binary Logistic Regression error = 3

USPS
USPS error rate = 9.37%
SGD-USPS
SGD-USPS one more epoch...
SGD-USPS error rate = 9.37%

```

**Comment by haifengl** [comment]
Thanks for quickly implementing this. It is interesting that the error rate is higher after one more epoch. Shall we shuffle the samples first? May be the samples are ordered by their class labels. It is not good for SGD.

File: core/src/main/java/smile/classification/LogisticRegression.java [hunk 1]
```diff
@@ -823,7 +834,78 @@ public double f(double[] w, double[] g) {
...
+				}
+			}
+		}
+	}	
+
+	/**
+	 * Tune the fixed learning rate for stochastic gradient descent
+	 * @param step usually quite small to avoid oscillation, default is 5e-5
+	 */
+    public void setStep(double step) {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/LogisticRegression.java]: `setLearningRate()` is better. THe learning rate in NeuralNetwork is called `eta`. Let's be consistent.

File: math/src/main/java/smile/math/Math.java [hunk 2]
```diff
@@ -4411,7 +4411,7 @@ public static double min(DifferentiableMultivariateFunction func, int m, double[
         // Current function value.
         double f = func.f(x, g);
 
-        logger.info(String.format("L-BFGS: initial function value: %.5g", f));
+        logger.debug(String.format("L-BFGS: initial function value: %.5g", f));
```
Comment (haifengl) [inline math/src/main/java/smile/math/Math.java]: Why `debug`? I prefer see the messages to asses if the algorithm is converging.

File: core/src/main/java/smile/classification/LogisticRegression.java [hunk 3]
```diff
@@ -75,7 +75,7 @@
  * 
  * @author Haifeng Li
  */
-public class LogisticRegression implements SoftClassifier<double[]>, Serializable {
+public class LogisticRegression implements SoftClassifier<double[]>, Serializable, OnlineClassifier<double[]> {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/LogisticRegression.java]: Please put `OnlineClassifier` before `Serializable`.

File: core/src/main/java/smile/classification/LogisticRegression.java (hunk 1 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/LogisticRegression.java]: so setLearningRate() or setEta()?

File: core/src/main/java/smile/classification/LogisticRegression.java (hunk 3 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/LogisticRegression.java]: consider it done.

File: math/src/main/java/smile/math/Math.java (hunk 2 above)
Comment (rayeaster) [reply math/src/main/java/smile/math/Math.java]: ok. I will revert this change.

**Comment by rayeaster** [comment]
"Shall we shuffle the samples first?"
any utility class I could use for sample shuffle? @haifengl

File: core/src/main/java/smile/classification/LogisticRegression.java (hunk 1 above)
Comment (haifengl) [reply core/src/main/java/smile/classification/LogisticRegression.java]: use `setLearningRate` for the method name and use `eta` for the variable name. Thanks.

**Comment by haifengl** [comment]
There is `permutate` method in `smile.math.Math`

File: core/src/main/java/smile/classification/LogisticRegression.java (hunk 1 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/LogisticRegression.java]: Thanks and got it.

**Comment by rayeaster** [comment]
@haifengl  code changed according your input and unit test result remains the same. Thanks,

File: core/src/main/java/smile/classification/LogisticRegression.java [hunk 4]
```diff
@@ -823,7 +834,78 @@ public double f(double[] w, double[] g) {
...
+		if (x.length != p) {
+			throw new IllegalArgumentException("Invalid input vector size: " + x.length);
+		}
+
+		if (k == 2) {
+			// calculate gradient for incoming data
+			double wx = dot(x, w);
+			double res = y - Math.logistic(wx);
+
+			double[] g = new double[p + 1];
```
Comment (haifengl) [inline core/src/main/java/smile/classification/LogisticRegression.java]: no need to create a new an array here. It introduces memory fragment and slows down the system when we update the model with a lot of samples. It is easy to combine this loop with the below one and just use a local variable.

File: core/src/main/java/smile/classification/LogisticRegression.java [hunk 5]
```diff
@@ -823,7 +834,78 @@ public double f(double[] w, double[] g) {
...
+			}
+		} else {
+			double[] prob = new double[k];
+			for (int j = 0; j < k; j++) {
+				prob[j] = dot(x, W[j]);
+			}
+
+			softmax(prob);
+
+			double[][] g = new double[k][p + 1];
```
Comment (haifengl) [inline core/src/main/java/smile/classification/LogisticRegression.java]: same as above comment.

**Comment by haifengl** [comment]
I think that the unit test with SGD should be revised. Note that our batch learning reaches the global optimal. By online learning with the same data, we won't get better results (actually could be worse). I suggest that we train with only part of data (says 80% samples) with batch algorithm. And then update the model with online learning with the rest of samples.

**Comment by rayeaster** [comment]
unit test results with 80% samples for batch algorithm, 20% training samples for sgd (shuffle enabled) : 
SGD-Segment error rate = 6.17%
SGD-Iris Binary Logistic Regression error = 8
SGD-USPS error rate = 9.57%

unit test results with optimal error rate for various parameter-tuning I tried : 
SGD-Segment error rate = **5.68**% （**70**% samples for batch algorithm）
SGD-Iris Binary Logistic Regression error = **3**（**95**% samples for batch algorithm）
SGD-USPS error rate = **9.57**%（**80**% samples for batch algorithm）

**Comment by haifengl** [comment]
Thanks!

**Comment by rayeaster** [comment]
my pleasure
